upgrade/v1beta1/upgrade.go

```diff
@@ -0,0 +1,210 @@
+
+	"github.com/jinzhu/copier"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	restclient "k8s.io/client-go/rest"
+
+	longhornV1beta1 "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+
+	"github.com/longhorn/longhorn-manager/upgrade/v1beta1/types"
+
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1beta1 to v1beta2:"
+)
+
+func UpgradeCRFromV1beta1ToV1beta2(config *restclient.Config, namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+" failed")
+	}()
+
+	lhClientV1beta1, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	if err := upgradeVolumes(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeEngineImages(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeBackupTargets(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeNodes(namespace, lhClientV1beta1, lhClient); err != nil {
+		return err
+	}
+
+	logrus.Infof("%v: completed", upgradeLogPrefix)
+	return nil
+}
+
+func copyConditions(fromConditions map[string]longhornV1beta1.Condition) ([]longhorn.Condition, error) {
+	toConditions := []longhorn.Condition{}
+	for _, fromCon := range fromConditions {
+		toCon := longhorn.Condition{}
+		if err := copier.Copy(&toCon, &fromCon); err != nil {
+			return nil, err
+		}
+		toConditions = append(toConditions, toCon)
+	}
+	return toConditions, nil
+}
+
+func CanUpgrade(config *restclient.Config, namespace string) (bool, error) {
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return false, errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	setting, err := lhClient.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameDefaultEngineImage), metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("setting %v not found", string(types.SettingNameDefaultEngineImage))
+			return true, nil
+		}
+
+		return false, errors.Wrap(err, fmt.Sprintf("unable to get setting %v", string(types.SettingNameDefaultEngineImage)))
+	}
+
+	return false, errors.Wrapf(err, "unable to upgrade from %v directly", setting.Value)
```

Comment:
Leaving a note here. Since like at v1alpha1 -> v1beta1, the longhorn-manager uses the `default-engine-image` Setting CR to determine whether it it v1alpha1 _or_ v1beta1, and the code is copy from there.

However, the code now is to handle upgrades from v1beta1 -> v1beta2, not sure should we keep this code logic here _or_ we could directly drop it.

Probably we could leave it as is, and see if there is any issue happens related to this one in the future.

Reply:
As I mentioned in this [comment](https://github.com/longhorn/longhorn-manager/pull/1136#discussion_r758306872), this code may confuse the contributors who are not familiar with the API version upgrade logic & compatibility. Since this function directly returns false without checking the setting. It's better to add some comments or explicitly check `setting` instead.